### PR TITLE
Fix stdout piped colored output on MacOS and Linux

### PR DESCRIPTION
**Pull request by varun-dc**
resolves https://github.com/dbt-labs/dbt-core/issues/4790

### Description

Prevent modifcation of stdout stream if the platform is not Windows. This ensures ANSI color codes in the output stream are correctly output to the stdout stream on Linux and MacOS. On Windows I noticed no difference, i.e. the code changes here seems to have no behavioral changes on Windows.

I'm not too familiar with the entirety of the dbt codebase nor the colorama codebase. But debugging the code, I noticed the stdout stream was a colorama object, not a the standard Python stdout stream. Which lead me to the code changes I've touched in this PR.

Regarding Windows, as I mentioned the behavior has not changed. But on Linux and MacOS, I'm seeing color output when I pipe and don't pipe out the stdout stream to another process (`dbt run` and `dbt run | cat`). This was the behavior previous; previous to this PR https://github.com/dbt-labs/dbt-core/pull/4474 I believe. On Windows, both `dbt run` and `dbt run | out-host -paging` (the equivalent of `| cat` afaik) behavior do the same thing as before (color output and no color output respectively). I'm not sure how to fix this on Windows.

To test this, I wasn't able to install my forked version using `pip install git+https://github.com/varun-dc/dbt-core@<branch name>`. So I instead modified the virtualenv pip installed copy locally to verify these changes. You can see the output of running `dbt run` and `dbt run | cat` here,
Screen Shot 2022-02-25 at 1 47 43 PM

### Checklist (**TODO**)

- [x] I have signed the [CLA](https://docs.getdbt.com/docs/contributor-license-agreements)
- [x] I have run this code in development and it appears to resolve the stated issue
- [x] This PR includes tests, or tests are not required/relevant for this PR
- [x] I have updated the `CHANGELOG.md` and added information about my change

### CODE REVIEW

File: core/dbt/logger.py [hunk 1]
```diff
@@ -20,20 +20,19 @@
...
 
-if sys.platform == "win32" and not os.getenv("TERM"):
-    colorama_wrap = False
-    colorama_stdout = colorama.AnsiToWin32(sys.stdout).stream
-
-elif sys.platform == "win32":
-    colorama_wrap = False
-
-colorama.init(wrap=colorama_wrap)
+    colorama.init(wrap=colorama_wrap)
```
Comment (ChenyuLInx) [inline core/dbt/logger.py:36]: This section is kind of hard to read before your change. I tried to reformat this section with the changes made, the logic is being updated to 
```
if sys.platform == "win32" and not os.getenv("TERM"):
    logging_stdout = colorama.AnsiToWin32(sys.stdout).stream
else:
    logging_stdout = sys.stdout

if sys.platform == "win32":
    colorama_wrap = False
    colorama.init(wrap=colorama_wrap)
```
Previous logic after similar refactor looks like
```
if sys.platform == "win32" and not os.getenv("TERM"):
    colorama_stdout = colorama.AnsiToWin32(sys.stdout).stream
else:
    colorama_stdout = sys.stdout

if sys.platform == "win32":
    colorama_wrap = False
else:
    colorama_wrap = True
colorama.init(wrap=colorama_wrap)
```
So seems like what you did was remove the initialization of `colorama ` when not running on windows. Is that what we wanted to do?

**Review by ChenyuLInx — COMMENTED** [review]
Hey @varun-dc  Thank you so much for taking a look at this and contributing!  Which system did you use to test this? I pulled the change and tested locally but it didn't solve the color issue for me. I am using an intel mac with zsh.

**Comment by varun-dc** [comment]
> Which system did you use to test this? I pulled the change and tested locally but it didn't solve the color issue for me. I am using an intel mac with zsh.

What I did was modify the dbt-core source in my Python virtualenv. I tried to install my fork via pip `git+https://github.com/varun-dc/dbt-core@fix-logging-color-pipe-stdout-windows`, but I couldn't get this to work. Hence modifying the venv locally.

I've just tested it again locally and can confirm I'm seeing correct behavior on MacOS 11.6 with zsh (I was using bash when I tested it before I made the PR).

Can you share your testing steps, I'm not too sure how you go about testing dbt internally. I can try to follow the same steps to help you troubleshoot any issues.

File: core/dbt/logger.py (hunk 1 above)
Comment (varun-dc) [reply core/dbt/logger.py:36]: Yes, my intent was to remove the initialization of `colarama` when not running on Windows as I noticed that seemed to resolve the issue for me when running on Mac and Linux

Regarding your suggested changes, yes that looks okay to me. I can try and test it myself locally to verify.

Comment (varun-dc) [reply core/dbt/logger.py:36]: Ideally we'd fix the issue on Windows. My solution here doesn't resolve the issue on Windows when I tested it.

But I'm not too familiar with Windows console or colorama to suggest any solutions at the moment. Any insight from the dbt dev team would be useful if possible.

**Comment by ChenyuLInx** [comment]
> hare your testing steps, I'm not too sure how you go about testing dbt internally. I can try to follow the same steps to help you troubleshoot any issues.

Hey @varun-dc , thanks for the fast response!!! This part of the code is new to me also. I was testing on mac 12.2.1 with a venv Python 3.9.10. I believe my testing step is the same as your's. I do a `gh pr checkout 4792` in `dbt-core` repo to checkout your changes locally. And the dbt cli I use is pointing to local code installed following [this](https://github.com/dbt-labs/dbt-core/blob/main/CONTRIBUTING.md#installation).  
And this is what I see on my terminal. I wonder if it is related to the python or the colorama version. I have`colorama==0.4.4`.

Screen Shot 2022-02-28 at 3 47 16 PM

**Comment by varun-dc** [comment]
> I pulled the change and tested locally but it didn't solve the color issue for me. I am using an intel mac with zsh.

Just a follow up, I think I'm able to reproduce this by installing my version locally 'properly'.
- `git clone https://github.com/varun-dc/dbt-core`
- `git checkout -t origin/fix-logging-color-pipe-stdout-windows` (I suppose this would effectively be equivalent to using the `gh pr checkout 4792` command you mentioned)
- enable a virtualenv (I use `python3 -m venv venv` and `source venv/bin/activate`)
- `cd dbt-core/core` (`cd` into the repo with my pr branch checked out)
- `pip install -e .`
- cd into a dbt project
- `dbt run | cat`

And doing this I don't see color as you saw as well. Let me debug and see what's happening and if my fix is actually correct or not.

File: core/dbt/events/functions.py [hunk 2]
```diff
@@ -55,19 +55,10 @@
...
 # converter. If the TERM var is set (as with Git Bash), then it's safe
 # to send escape characters and no log handler injection is needed.
-colorama_stdout = sys.stdout
-colorama_wrap = True
-
-colorama.init(wrap=colorama_wrap)
-
 if sys.platform == "win32" and not os.getenv("TERM"):
-    colorama_wrap = False
-    colorama_stdout = colorama.AnsiToWin32(sys.stdout).stream
```
Comment (varun-dc) [inline core/dbt/events/functions.py:65]: Unless this statement has a side-effect or is being used externally, it appears to be deadcode, hence I removed it. Doing a grep over the `dbt-core` repo, I see no references outside of this file for `colorama_stdout`.

Comment (ChenyuLInx) [reply core/dbt/events/functions.py:65]: @nathaniel-may Seems like you added this section

File: core/dbt/logger.py (hunk 1 above)
Comment (ChenyuLInx) [reply core/dbt/logger.py:28]: I personally don't have much experience with windows either. We can leave the windows part out!

**Comment by ChenyuLInx** [comment]
> > I pulled the change and tested locally but it didn't solve the color issue for me. I am using an intel mac with zsh.
> 
> Just a follow up, I think I'm able to reproduce this by installing my version locally 'properly'.
> 
> * `git clone https://github.com/varun-dc/dbt-core`
> * `git checkout -t origin/fix-logging-color-pipe-stdout-windows` (I suppose this would effectively be equivalent to using the `gh pr checkout 4792` command you mentioned)
> * enable a virtualenv (I use `python3 -m venv venv` and `source venv/bin/activate`)
> * `cd dbt-core/core` (`cd` into the repo with my pr branch checked out)
> * `pip install -e .`
> * cd into a dbt project
> * `dbt run | cat`
> 
> And doing this I don't see color as you saw as well. Let me debug and see what's happening and if my fix is actually correct or not.

@varun-dc  We really appreciate the efforts here!!! I think these are the exact steps! It's interesting that this and directly modifying the file produce different results

File: core/dbt/events/functions.py [hunk 3]
```diff
@@ -55,19 +55,8 @@
...
-if sys.platform == "win32" and not os.getenv("TERM"):
-    colorama_wrap = False
-    colorama_stdout = colorama.AnsiToWin32(sys.stdout).stream
-
-elif sys.platform == "win32":
-    colorama_wrap = False
-
-colorama.init(wrap=colorama_wrap)
+if sys.platform == "win32":
+    colorama.init(wrap=False)
```
Comment (varun-dc) [inline core/dbt/events/functions.py:59]: I think this logic is all we need here.

File: core/dbt/logger.py [hunk 4]
```diff
@@ -20,20 +20,11 @@
...
-
-elif sys.platform == "win32":
-    colorama_wrap = False
-
-colorama.init(wrap=colorama_wrap)
+logging_stdout = sys.stdout
+if sys.platform == "win32":
+    if not os.getenv("TERM"):
+        logging_stdout = colorama.AnsiToWin32(sys.stdout).stream
+    colorama.init(wrap=False)
```
Comment (varun-dc) [inline core/dbt/logger.py:27]: And I think this logic is the minimal reduction of the logic we had before. Again, please correct any logical errors in this refactoring

**Review by varun-dc — COMMENTED** [review]
I've updated the code now. Had to fix up some code in `functions.py`. I've tested it by installing the modified version of `dbt-core` using `pip`. It looks good for me on Linux and MacOS.

File: CHANGELOG.md [hunk 5]
```diff
@@ -37,6 +37,7 @@ Contributors:
 - Fix bug causing empty node level meta, snapshot config errors ([#4459](https://github.com/dbt-labs/dbt-core/issues/4459), [#4726](https://github.com/dbt-labs/dbt-core/pull/4726))
 - Fix slow `dbt run` when using Postgres adapter, by deduplicating relations in `postgres_get_relations` ([#3058](https://github.com/dbt-labs/dbt-core/issues/3058), [#4521](https://github.com/dbt-labs/dbt-core/pull/4521))
 - Fix partial parsing bug with multiple snapshot blocks ([#4771](https//github.com/dbt-labs/dbt-core/issues/4772), [#4773](https://github.com/dbt-labs/dbt-core/pull/4773))
+- Fix lack of color output on Linux and MacOS when piping the output into another process using the shell pipe (`|`) [#4792](https://github.com/dbt-labs/dbt-core/pull/4792)
```
Comment (ChenyuLInx) [inline CHANGELOG.md:41]:
```suggestion
Contributors:
* [@varun-dc ](https://github.com/varun-dc) ([#4792](https://github.com/dbt-labs/dbt-core/pull/4792))

**Review by ChenyuLInx — CHANGES_REQUESTED** [review]
Thank you so much for working through this!! Don't forget to add yourself to the Contributors!

File: CHANGELOG.md (hunk 5 above)
Comment (varun-dc) [reply CHANGELOG.md]: Thanks

**Review by ChenyuLInx — APPROVED** [review]

**Comment by ChenyuLInx** [comment]
@varun-dc  Congrats on the first PR!!! Thanks for working on this! We are putting it targeting 1.1 release.

**Pull request merged**